File: src/main/java/com/example/the_machine/service/RunPolicy.java
package com.example.the_machine.service;

/**
 * Policy configuration for run execution limits and retry behavior.
 */
public record RunPolicy(int maxTriesPerOp, double maxCostUsd, int maxDepth, int maxNodes) {

  public RunPolicy {
    if (maxTriesPerOp <= 0) {
      throw new IllegalArgumentException("maxTriesPerOp must be positive");
    }
    if (maxCostUsd <= 0) {
      throw new IllegalArgumentException("maxCostUsd must be positive");
    }
    if (maxDepth <= 0) {
      throw new IllegalArgumentException("maxDepth must be positive");
    }
    if (maxNodes <= 0) {
      throw new IllegalArgumentException("maxNodes must be positive");
    }
  }

  public static RunPolicy defaultPolicy() {
    return new RunPolicy(3, 10.0, 10, 100);
  }
}

File: src/main/kotlin/com/example/the_machine/service/RunPolicy.kt
package com.example.the_machine.service

/**
 * Policy configuration for run execution limits and retry behavior.
 */
data class RunPolicy(
  val maxTriesPerOp: Int,
  val maxCostUsd: Double,
  val maxDepth: Int,
  val maxNodes: Int
) {

  init {
    require(maxTriesPerOp > 0) { "maxTriesPerOp must be positive" }
    require(maxCostUsd > 0) { "maxCostUsd must be positive" }
    require(maxDepth > 0) { "maxDepth must be positive" }
    require(maxNodes > 0) { "maxNodes must be positive" }
  }

  companion object {

    fun defaultPolicy(): RunPolicy {
      return RunPolicy(3, 10.0, 10, 100)
    }
  }
}